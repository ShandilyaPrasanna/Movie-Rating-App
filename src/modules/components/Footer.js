import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";

// Footer Component
const useStyles = makeStyles(theme => ({
  root: {
    display: "flex",
    flexDirection: "column",
    minHeight: "73vh"
  },
  footer: {
    padding: theme.spacing(2),
    marginTop: "auto",
    fontSize: "18px",
    backgroundColor: "transparent"
  }
}));

export default function Footer() {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <footer className={classes.footer}>
        <Container maxWidth="sm">
          Follow us on social media Sites :- <a>Facebook</a> <a>Twitter</a>{" "}
          <a>Instagram</a> <a>Linkedln</a>
        </Container>
      </footer>
    </div>
  );
}
