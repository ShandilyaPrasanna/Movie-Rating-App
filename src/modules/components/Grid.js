import React from "react";
import { connect } from "react-redux";
import compose from "compose-function";

//Material ui components
import Card from "@material-ui/core/Card";
import Grid from "@material-ui/core/Grid";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import Typography from "@material-ui/core/Typography";
import { withStyles } from "@material-ui/core/styles";
import Container from "@material-ui/core/Container";

//Icons
import Grade from "@material-ui/icons/Grade";
import ClapperImage from "../../assets/img/clapperboard-icon.jpg";

const styles = () => ({
  card: {
    height: "100%",
    display: "flex",
    flexDirection: "column"
  },
  title: {
    font: "30px Arial,sans-serif",
    fontWeight: "normal",
    lineHeight: "110%"
  },
  gradeIcon: { marginTop: "-4px", color: "gold" },
  cardContent: {
    display: "flex",
    flexDirection: "column",
    marginTop: "15px"
  },
  cardMedia: { width: "30%", marginLeft: "auto" }
});

//Component display Movie details in grid based on store state change
export class movieGrid extends React.Component {
  displayRatings = reviews => {
    const { classes } = this.props;
    return (
      <CardActions>
        {reviews.map((rating, index) => {
          const style = { padding: "3px" };
          if (index !== 0) {
            style.marginLeft = "auto";
          }
          return (
            <React.Fragment>
              <Typography
                variant="body2"
                color="primary"
                component="p"
                style={style}
              >
                {rating.rating}
              </Typography>
              <Grade className={classes.gradeIcon} />
              <Typography variant="body2" color="primary" component="p">
                {rating.provider}
              </Typography>
            </React.Fragment>
          );
        })}
      </CardActions>
    );
  };

  displayMovieDetails = (type, value) => (
    <Container style={{ display: "flex" }}>
      <Typography variant="h8" component="h3">
        {`${type}`}
      </Typography>
      <Typography
        variant="body2"
        color="primary"
        component="p"
        style={{ padding: "3px" }}
      >
        {`${value}`}
      </Typography>
    </Container>
  );
  render() {
    const { classes, movies } = this.props;
    return (
      <Grid container spacing={2} style={{ marginTop: "20px" }}>
        {movies.map(movie => (
          <Grid item key={movie.id} xs={12} sm={6} md={6}>
            <Card className={classes.card}>
              <CardContent>
                <Typography
                  variant="h5"
                  component="h2"
                  className={classes.title}
                >
                  {`${movie.title} (${movie.year})`}
                </Typography>
                <CardContent style={{ display: "flex" }}>
                  <CardContent className={classes.cardContent}>
                    {this.displayMovieDetails("Genre:", movie.genre)}
                    {this.displayMovieDetails("Duration:", movie.duration)}
                    {this.displayMovieDetails("Director:", movie.director)}
                  </CardContent>
                  <CardMedia
                    component="img"
                    alt="Contemplative Reptile"
                    height="140"
                    image={ClapperImage}
                    className={classes.cardMedia}
                    title="Watch Movie"
                  />
                </CardContent>
              </CardContent>
              <CardContent>
                <Typography variant="body2" color="textSecondary" component="p">
                  {movie.description}
                </Typography>
              </CardContent>
              {this.displayRatings(movie.reviews)}
            </Card>
          </Grid>
        ))}
      </Grid>
    );
  }
}

export function mapStateToProps(state) {
  return {
    movies: state.app.movies || []
  };
}

const enhance = compose(
  withStyles(styles),
  connect(mapStateToProps)
);

export default enhance(movieGrid);
