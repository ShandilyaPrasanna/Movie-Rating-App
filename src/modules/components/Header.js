import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import IconButton from "@material-ui/core/IconButton";
import logo from "../../assets/img/logo.png";

//Header Component
const useStyles = makeStyles(theme => ({
  logo: {
    marginLeft: "50%",
    transform: "translateX(-50%)"
  }
}));

export default function Header() {
  const classes = useStyles();
  return (
    <div style={{ margin: "30px" }}>
      <img src={logo} height="50px" width="200px" className={classes.logo} />
    </div>
  );
}
