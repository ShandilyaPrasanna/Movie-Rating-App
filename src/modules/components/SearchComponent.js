import React from "react";

//utils
import { withStyles } from "@material-ui/core/styles";
import { debounce } from "lodash";
import compose from "compose-function";
import { connect } from "react-redux";

//Material ui components
import Paper from "@material-ui/core/Paper";
import Container from "@material-ui/core/Container";
import InputBase from "@material-ui/core/InputBase";
import IconButton from "@material-ui/core/IconButton";
import TextField from "@material-ui/core/TextField";
import Checkbox from "@material-ui/core/Checkbox";
import Button from "@material-ui/core/Button";

//Material ui icons
import MenuIcon from "@material-ui/icons/Menu";
import SearchIcon from "@material-ui/icons/Search";

//actions
import { getMoviesAction, isLoading } from "../actions";

const styles = () => ({
  root: {
    padding: "2px 4px",
    display: "flex",
    alignItems: "center",
    width: "100%"
  },
  input: {
    marginLeft: 8,
    flex: 1
  },
  filterContainer: { display: "flex", marginTop: "22px" }
});

//Components handles search and filter functionality
export class SearchComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      checked: { year: false, genre: false },
      filterValues: { year: 2019, genre: "" },
      searchTxt: ""
    };
  }

  componentWillUnmount() {
    this.searchHandler.cancel();
  }

  makeApiRequest = value => {
    const searchTxt = value || this.state.searchTxt;
    const { getMoviesAction, isLoading } = this.props;
    const { filterValues, checked } = this.state;
    const filter = [];
    if (checked.year) {
      filter.push({ year: filterValues.year });
    }
    if (checked.genre) {
      filter.push({ genre: filterValues.genre });
    }
    const endPoint = `search?key=${searchTxt}${
      filter.length !== 0 ? `&filter=${JSON.stringify(filter)}` : ``
    }`;

    getMoviesAction(endPoint);
    isLoading(true);
  };

  searchHandler = debounce(searchTxt => {
    this.setState({ searchTxt }, () => {
      this.makeApiRequest(this.state.searchTxt);
    });
  }, 500);

  handleFilterValues = event => {
    const { filterValues } = this.state;
    const { name, value } = event.target;
    this.setState({ filterValues: { ...filterValues, [name]: value } });
  };

  displayFilter = () => {
    const { checked } = this.state;
    const visibility = checked.genre ? "visible" : "hidden";
    const { classes } = this.props;
    return (
      <Container className={classes.filterContainer} maxWidth="sm">
        <div>
          Filter By:- Year
          <Checkbox
            id="checkboxYear"
            checked={checked.year}
            style={{
              color: "#ffff"
            }}
            onChange={() => {
              this.setState({ checked: { ...checked, year: !checked.year } });
            }}
          />
          {checked.year && (
            <TextField
              id="standard-number"
              name="year"
              defaultValue={2019}
              style={{
                backgroundColor: "#fff",
                width: "100px",
                marginTop: "6px"
              }}
              type="number"
              onChange={this.handleFilterValues}
              InputProps={{
                disableUnderline: true,
                style: { marginLeft: "30px" }
              }}
            />
          )}
        </div>
        <div style={{ marginLeft: "auto" }}>
          Genre
          <Checkbox
            id="checkboxGenre"
            checked={checked.genre}
            style={{
              color: "#ffff"
            }}
            onChange={() => {
              this.setState({ checked: { ...checked, genre: !checked.genre } });
            }}
          />
          <TextField
            id="standard-genre"
            name="genre"
            style={{
              backgroundColor: "#fff",
              width: "100px",
              visibility,
              marginTop: "6px"
            }}
            onChange={this.handleFilterValues}
            InputProps={{
              disableUnderline: true,
              style: { marginLeft: "30px" }
            }}
          />
        </div>
        <Button
          id="applyFiler"
          variant="contained"
          style={{ marginLeft: "25px" }}
          onClick={() => {
            this.makeApiRequest();
          }}
        >
          Apply
        </Button>
      </Container>
    );
  };

  render() {
    const { classes } = this.props;
    return (
      <React.Fragment>
        <Container maxWidth="sm" className={classes.root}>
          <Paper className={classes.root}>
            <IconButton>
              <MenuIcon />
            </IconButton>
            <InputBase
              id="searchBar"
              className={classes.input}
              placeholder="Search Movies Here"
              onChange={event => {
                this.searchHandler(event.target.value);
              }}
            />
            <IconButton>
              <SearchIcon />
            </IconButton>
          </Paper>
        </Container>
        {this.displayFilter()}
      </React.Fragment>
    );
  }
}

const enhance = compose(
  withStyles(styles),
  connect(
    null,
    {
      getMoviesAction,
      isLoading
    }
  )
);

export default enhance(SearchComponent);
