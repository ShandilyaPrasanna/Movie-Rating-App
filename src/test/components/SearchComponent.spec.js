import React from "react";
import ReactDOM from "react-dom";
import toJson from "enzyme-to-json";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { SearchComponent } from "../../modules/components/SearchComponent";

jest.mock("react-dom");

configure({ adapter: new Adapter() });
describe("SearchComponent Test", () => {
  let tree;
  const getMoviesAction = jest.fn();
  const isLoading = jest.fn();

  const props = {
    getMoviesAction,
    isLoading,
    classes: {}
  };
  beforeEach(() => {
    tree = shallow(<SearchComponent {...props} />);
  });

  it("renders SearchComponent ", () => {
    expect(toJson(tree)).toMatchSnapshot();
  });
  it("check onChange of searchBar ", () => {
    tree.find("#searchBar").simulate("change", { target: { value: "apple" } });
    setTimeout(function() {
      try {
        done();
      } catch (e) {
        done.fail(e);
      }
      expect(tree.state("searchTxt")).toBe("apple");
    }, 500);
  });
  it("check checkbox state change ", () => {
    tree.find("#checkboxGenre").simulate("change");
    expect(tree.state("checked").genre).toBe(true);
  });
  it("check checkbox state change ", () => {
    tree.find("#checkboxYear").simulate("change");
    expect(tree.state("checked").year).toBe(true);
  });
  it("Apply filter button functionality ", () => {
    tree.setState({ checked: { year: true, genre: true } });
    tree.find("#applyFiler").simulate("click");
    expect(getMoviesAction).toHaveBeenCalled();
    expect(isLoading).toHaveBeenCalled();
  });

  it(" filter year change handler ", () => {
    tree
      .find("#standard-genre")
      .simulate("change", { target: { value: 2019 } });
    expect(tree.state("filterValues").year).toBe(2019);
  });
});
