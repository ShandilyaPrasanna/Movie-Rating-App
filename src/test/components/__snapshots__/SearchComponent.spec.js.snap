// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent Test renders SearchComponent  1`] = `
<Fragment>
  <WithStyles(ForwardRef(Container))
    maxWidth="sm"
  >
    <WithStyles(ForwardRef(Paper))>
      <WithStyles(ForwardRef(IconButton))>
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(InputBase))
        id="searchBar"
        onChange={[Function]}
        placeholder="Search Movies Here"
      />
      <WithStyles(ForwardRef(IconButton))>
        <SearchIcon />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Container))
    maxWidth="sm"
  >
    <div>
      Filter By:- Year
      <WithStyles(ForwardRef(Checkbox))
        checked={false}
        id="checkboxYear"
        onChange={[Function]}
        style={
          Object {
            "color": "#ffff",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      Genre
      <WithStyles(ForwardRef(Checkbox))
        checked={false}
        id="checkboxGenre"
        onChange={[Function]}
        style={
          Object {
            "color": "#ffff",
          }
        }
      />
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "disableUnderline": true,
            "style": Object {
              "marginLeft": "30px",
            },
          }
        }
        id="standard-genre"
        name="genre"
        onChange={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "marginTop": "6px",
            "visibility": "hidden",
            "width": "100px",
          }
        }
      />
    </div>
    <WithStyles(ForwardRef(Button))
      id="applyFiler"
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "25px",
        }
      }
      variant="contained"
    >
      Apply
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Container))>
</Fragment>
`;
